README.md
# Webp Converter and Analytics

![](https://img.shields.io/badge/webp-batch%20converter-orange.svg)
![](https://img.shields.io/badge/webp-batch%20analytics-orange.svg)
![](https://img.shields.io/badge/license-Apache2-blue.svg)
[![](https://img.shields.io/badge/readme-English-blue.svg)](https://github.com/Jacksgong/webp-converter)
[![](https://img.shields.io/badge/readme-中文-blue.svg)](https://github.com/Jacksgong/webp-converter/blob/master/README-zh.md)
[![](https://img.shields.io/badge/pip-v4.1.0%20webp--converter-yellow.svg)](https://pypi.python.org/pypi/webp-converter)

An powerful webp batch converter and differ analytics tool.

You can use this tool to converter batch images(png/jpg) to webp and output size changes.

![](https://github.com/Jacksgong/webp-converter/raw/master/arts/webp-converter.png)


## Purpose

1. Convert batch images from PNG/JPG format to Webp format
2. WILL NOT convert images if its webp one is larger than origin one automatically(the origin one will be copied to `[output-directory]/origin/` directory)
3. WILL NOT convert images if it is failed to convert to webp one automatically(the origin image of failed one will be copied to `[output-directory]/failed` directory)
4. You can ignore all images which has transparency pixel if you want with `--ignore-transparency-image` config(the origin image of has-transparency-image will be copied to `/webp-converted/transparency` directory)
5. Output convert result, like how much size reduces, how many files skip convert, etc...
6. You can replace the images with converted-webp image automatically if you want with `replace: true` config

## Install

```shell
brew install webp
pip install webp-converter
```

## Use

![](https://github.com/Jacksgong/webp-converter/raw/master/arts/help.png)

#### Example

If you just want to convert all images on the current files when it can be smaller after converted and **replace** the origin one:

```shell
webpc --r
```

![](https://github.com/Jacksgong/webp-converter/raw/master/arts/demo-1.png)


If you just want to convert all images on the current folder and with **quality-ratio** to `70` and **output** to `~/Downloads/test-converted` when it can be smaller after converted:

```shell
webpc -q=70 -o=~/Downloads/test-converted/
```

![](https://github.com/Jacksgong/webp-converter/raw/master/arts/demo-2.png)


If you just want to convert all images on the current folder and with **clean** the `webp-converted` folder if it exist and with **quality-ratio** to `95` and **output** to `./webp-converted` and **ignore images when it has transparency** on it when it can be smaller after converted:

```shell
webpc --c --ignore-transparency-image -q=95
```
![](https://github.com/Jacksgong/webp-converter/raw/master/arts/demo-3.png)

If you want to convert ignore filename contain the sepcial word, you can use `--ignore-filename-match`, the following demo is skip convert all files which name contains '.9':

```shell
webpc --r --ignore-filename-match='.9'
```

If you want to convert all **images on** `~/Downloads/img/test` folder and **output** converted result to `~/Downloads/test-converted` folder and only converted origin images when it isn't has **same name** `.webp` file on `~/Downloads/test-converted` folder(just not with `--c` argument) when it can be smaller after converted:


```shell
webpc -o=~/Downloads/test-converted/ ~/Downloads/img/test
```

![](https://github.com/Jacksgong/webp-converter/raw/master/arts/demo-4.png)

## My Terminal Config

If you want to adapter the same theme like screenshot above, it's very easy:

- Firstly, please use [powerlevel9k](https://github.com/bhilburn/powerlevel9k) theme(Install the Powerlevel9k Theme and Powerline Fonts as the powerlevel9k repo readme doc said).
- Secondly, please config the [iTerm2-Neutron](https://github.com/Ch4s3/iTerm2-Neutron) color scheme.
- Thirdly, please config your shell(If you are using zsh, just add following code to the `~/.zshrc` file):
```
POWERLEVEL9K_LEFT_PROMPT_ELEMENTS=(dir vcs)
POWERLEVEL9K_RIGHT_PROMPT_ELEMENTS=(status time)
POWERLEVEL9K_TIME_FORMAT="%D{%H:%M:%S}"
POWERLEVEL9K_NODE_VERSION_BACKGROUND='022'
POWERLEVEL9K_SHORTEN_DIR_LENGTH=2
```

## License

```
Copyright 2017, JacksGong(https://jacksgong.com)
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
  http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# CHANGE LOG

## V4.1.0

2021.01.19

- Support ignore files which name contain special word use `--ignore-filename-match`
- Support Python3
- Fix remove temp file issue when diff size is 0

## V4.0.1

2017.09.06

- Handle the case of interrupt by user

## V4.0.0

2017.09.05

- Convert to pip package.

## V3.0.0

- Output consume time
- Support replace the origin image with the converted webp image automatically: config with `replace: true` on `.webp.conf`

## V2.0.0

- Support ignore images which has transparency pixel: config with `ignore-transparency-image: true` on `.webp.conf`

## V1.0.0

- init release.


webpc/helper.py
#!/usr/bin/python -u

"""
Copyright 2017, JacksGong(https://jacksgong.com)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

import sys
from os import makedirs, environ
from os.path import getsize, join, exists
from shutil import copyfile

import re

RESET = '\033[0m'
BLACK, RED, GREEN, YELLOW, BLUE, MAGENTA, CYAN, WHITE = range(8)

NO_HOME_PATH = re.compile(r'~/(.*)')
HOME_PATH = environ['HOME']


# get the home case path
def handle_home_case(path):
    path = path.strip()
    if path.startswith('~/'):
        path = HOME_PATH + '/' + NO_HOME_PATH.match(path).groups()[0]
    return path


def print_blue(msg):
    print(colorize(msg, fg=BLUE))


def print_warn(msg):
    print(colorize(msg, fg=YELLOW))


def termcolor(fg=None, bg=None):
    codes = []
    if fg is not None: codes.append('3%d' % fg)
    if bg is not None: codes.append('10%d' % bg)
    return '\033[%sm' % ';'.join(codes) if codes else ''


def colorize(message, fg=None, bg=None):
    return termcolor(fg, bg) + message + RESET


def print_process(message):
    print(colorize(message, fg=YELLOW))


def size_diff(left, right):
    return getsize(left) - getsize(right)


def human_bytes(B):
    """Return the given bytes as a human friendly KB, MB, GB, or TB string"""
    B = float(B)
    KB = float(1024)
    MB = float(KB ** 2)  # 1,048,576
    GB = float(KB ** 3)  # 1,073,741,824
    TB = float(KB ** 4)  # 1,099,511,627,776

    if B < KB:
        return '{0} {1}'.format(B, 'Bytes' if 0 == B > 1 else 'Byte')
    elif KB <= B < MB:
        return '{0:.2f} KB'.format(B / KB)
    elif MB <= B < GB:
        return '{0:.2f} MB'.format(B / MB)
    elif GB <= B < TB:
        return '{0:.2f} GB'.format(B / GB)
    elif TB <= B:
        return '{0:.2f} TB'.format(B / TB)


def resource_path(relative_path):
    # noinspection PyProtectedMember
    return join(sys._MEIPASS, relative_path) if hasattr(sys, '_MEIPASS') else relative_path


def copyfile_safe(input_file_path, target_file_dir, file_name):
    if not exists(target_file_dir):
        makedirs(target_file_dir)

    target_file_path = target_file_dir + file_name
    if exists(target_file_path):
        print_warn(
            'we will not copy file[%s] because of there is a another file with the same name has been handled.' % input_file_path)
    else:
        copyfile(input_file_path, target_file_dir + file_name)


webpc/__init__.py
#!/usr/bin/python -u

"""
Copyright 2017, JacksGong(https://jacksgong.com)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""
import argparse
import os
import time
from os import makedirs, remove
from os.path import exists, join
from shutil import rmtree
from sys import exit

from webpc.converter import Converter, RESULT_ALREADY_EXIST, RESULT_FAILED, RESULT_WEBP_LARGER, RESULT_WITH_TRANSPARENCY
from webpc.helper import print_blue, resource_path, colorize, CYAN, GREEN, human_bytes, copyfile_safe, handle_home_case, \
    print_warn

__version__ = '4.1.0'
__author__ = 'JacksGong'

print("-------------------------------------------------------")
print("               Webp Converter v" + __version__)
print("")
print("Thanks for using okcat! Now, the doc is available on: ")
print_blue("      https://github.com/Jacksgong/webp-converter")
print("")
print("                      Have Fun!")
print("-------------------------------------------------------")


def main():
    default_output_directory = "%s/%s" % (os.getcwd(), 'webp-converted/')
    parser = argparse.ArgumentParser(description='Converter and analytics batch of images(png/jpg) to webp')
    parser.add_argument('image_directory', nargs='*', default=[os.getcwd()], help='Origin images directory path')
    parser.add_argument('-q', '--quality-ratio', dest='quality_ratio', type=int, default=100,
                        help='Quality ratio, between 0 to 100, 100 is lossless, 0 is highest compression ratio, default value is 100')
    parser.add_argument('-o', '--output-directory', dest='output_directory',
                        default=default_output_directory,
                        help='the output directory of converted path, default value is ./webp-converted/')
    parser.add_argument('--ignore-transparency-image', dest='ignore_transparency_image', action='store_true',
                        help='Whether need to ignore images which has transparency pixel on it, default value is false')
    parser.add_argument('--r', dest='replace_origin', action='store_true',
                        help='Whether replace the origin images files directly or not, default value is false')
    parser.add_argument('--c', dest='clear_env', action='store_true',
                        help='Whether need to clean the output directory before convert images, default value is false')
    parser.add_argument('--debug', dest='debug', action='store_true',
                        help='Whether need to print debug message, default value is false')
    parser.add_argument('--ignore-filename-match', dest='ignore_filename_match', default='',
                        help='ignore file when the name contain the provide value(such as .9)')

    args = parser.parse_args()

    start_time = time.time()

    input_directory = handle_home_case(args.image_directory[0])
    output_directory = handle_home_case(args.output_directory)
    quality_ratio = args.quality_ratio
    ignore_transparency_image = args.ignore_transparency_image
    replace_origin = args.replace_origin
    clean_env = args.clear_env
    is_debug = args.debug
    ignore_filename_match = args.ignore_filename_match

    if replace_origin and (clean_env or output_directory != default_output_directory):
        exit(
            "the --c and -o can't company with --r, because of --r means you need to store the converted images on the origin "
            "images directory and replace the origin images directly, and -o is means you want to store the converted images "
            "on some different directory, and --c is means you want to clear the output directory you provided with -o before"
            " convert.")

    if replace_origin:
        print('origin images will be replace with webp images on ' + input_directory)
    else:
        print('origin images on ' + input_directory + ' will be converted to ' + output_directory)

    command_prefix = '%s ' % resource_path('cwebp')
    if quality_ratio >= 100:
        command_prefix += '-lossless -q 100 '
    else:
        command_prefix += '-q ' + quality_ratio.__str__() + ' '

    keep_origin_path = output_directory + 'origin/'
    convert_fail_path = output_directory + 'failed/'
    transparency_image_path = output_directory + 'transparency/'
    swap_webp_path = join(output_directory, 'swap.webp') if not replace_origin else join(input_directory, 'swap.webp')

    if clean_env and exists(output_directory):
        rmtree(output_directory)
        print("clear all env on " + output_directory)

    if not replace_origin and not exists(output_directory):
        makedirs(output_directory)

    if exists(swap_webp_path):
        remove(swap_webp_path)

    output_data = OutputData()
    try:
        loop(
            input_directory=input_directory,
            output_directory=output_directory,
            swap_webp_path=swap_webp_path,
            transparency_image_path=transparency_image_path,
            quality_ratio=quality_ratio,
            convert_fail_path=convert_fail_path,
            keep_origin_path=keep_origin_path,
            ignore_transparency_image=ignore_transparency_image,
            is_debug=is_debug,
            replace_origin=replace_origin,
            output_data=output_data,
            ignore_filename_match=ignore_filename_match
        )
    except KeyboardInterrupt:
        print('')
        print('INTERRUPT BY USER.')
        if exists(swap_webp_path):
            remove(swap_webp_path)
        pass

    output_data.dump(
        replace_origin=replace_origin,
        input_directory=input_directory,
        output_directory=output_directory,
        start_time=start_time,
        ignore_filename_match=ignore_filename_match
    )


class OutputData:
    all_reduce_size = 0
    valid_convert_file_count = 0
    failed_convert_count = 0
    scan_file_count = 0
    skip_file_count = 0
    skip_transparency_file_count = 0
    skip_ignore_match_name_file_count = 0

    def __init__(self):
        pass

    def dump(self, replace_origin, input_directory, output_directory, start_time, ignore_filename_match):
        print('-----------------------------------------------')
        print(' ')
        if replace_origin:
            colorize('Replace %s image files on %s' % (self.valid_convert_file_count, input_directory),
                     fg=CYAN)
        else:
            print(get_result('All files handled on: ', output_directory, fg=CYAN))

        print(get_result('Consume: ', (time.time() - start_time), _format='%s%.3fs', fg=CYAN))
        print(' ')
        print(get_result('Scan files count: ', self.scan_file_count))
        print(get_result('Converted files count: ', self.valid_convert_file_count))
        print(get_result('Reduce size: ', human_bytes(self.all_reduce_size)))
        print(get_result('Skip files(because convert failed) count: ', self.failed_convert_count))
        print(get_result('Skip files(because the webp one is greater than origin one) count: ', self.skip_file_count))
        if self.skip_transparency_file_count > 0:
            print(get_result('Skip files(because there is transparency) count: ', self.skip_transparency_file_count))
        if self.skip_ignore_match_name_file_count > 0:
            print(get_result('Skip files(because their name contain %s) count: ' % ignore_filename_match,
                             self.skip_ignore_match_name_file_count))
        print (' ')
        print ('-----------------------------------------------')


def get_result(title, message, _format='%s%s', fg=GREEN):
    return _format % (colorize(title, fg=fg), message)


def loop(input_directory, output_directory, swap_webp_path, transparency_image_path, quality_ratio, convert_fail_path,
         keep_origin_path, ignore_transparency_image, is_debug,
         replace_origin, output_data, ignore_filename_match):
    handler = Converter(swap_webp_path, quality_ratio)
    for subdir, dirs, files in os.walk(input_directory):
        for file_name in files:
            if file_name == '.DS_Store':
                continue
            if ignore_filename_match != '' and file_name.__contains__(ignore_filename_match):
                if is_debug:
                    print('ignore %s because it contain %s' % (file_name, ignore_filename_match))
                output_data.skip_ignore_match_name_file_count += 1
                continue
            if not file_name.endswith('.jpg') and not file_name.endswith('.png'):
                continue

            if output_directory in subdir:
                continue

            output_data.scan_file_count += 1
            input_file_path = join(subdir, file_name)
            output_file_name = file_name.rsplit('.', 1)[0] + '.webp'
            if replace_origin:
                output_file_path = join(subdir, output_file_name)
            else:
                output_file_path = join(output_directory, output_file_name)

            result, reduce_size = handler.convert(ignore_transparency_image, input_file_path, file_name,
                                                  output_file_path, is_debug)
            if reduce_size > 0:
                output_data.all_reduce_size += reduce_size
                if result != RESULT_ALREADY_EXIST:
                    output_data.valid_convert_file_count += 1
            elif result == RESULT_FAILED:
                output_data.failed_convert_count += 1
            elif result == RESULT_WEBP_LARGER:
                output_data.skip_file_count += 1
            elif result == RESULT_WITH_TRANSPARENCY:
                output_data.skip_transparency_file_count += 1

            if replace_origin:
                if reduce_size > 0:
                    remove(input_file_path)
            else:
                if result == RESULT_WITH_TRANSPARENCY:
                    copyfile_safe(input_file_path, transparency_image_path, file_name)
                elif result == RESULT_FAILED:
                    copyfile_safe(input_file_path, convert_fail_path, file_name)
                elif result == RESULT_WEBP_LARGER:
                    copyfile_safe(input_file_path, keep_origin_path, file_name)


webpc/converter.py
#!/usr/bin/python -u

"""
Copyright 2017, JacksGong(https://jacksgong.com)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""
import subprocess
import os
from os import remove, rename
from os.path import exists

from webpc.helper import size_diff, print_process, resource_path

RESULT_SUCCESS = 0
RESULT_FAILED = -1
RESULT_WEBP_LARGER = -2
RESULT_WITH_TRANSPARENCY = -3
RESULT_ALREADY_EXIST = -4

FNULL = open(os.devnull, 'w')


class Converter:
    swap_webp_path = None
    command_prefix = list()

    def __init__(self, swap_webp_path, quality_ratio):
        self.swap_webp_path = swap_webp_path

        command = list()
        command.append(resource_path("cwebp"))
        # command_prefix = '%s ' % resource_path('cwebp')
        if quality_ratio >= 100:
            command.append("-lossless")
            command.extend(['-q', '100'])
            # command_prefix += '-lossless -q 100 '
        else:
            command.extend(['-q', quality_ratio.__str__()])
            # command_prefix += '-q ' + quality_ratio.__str__() + ' '

        self.command_prefix = command

    def convert(self, ignore_transparency_img, image_file_path, image_file_name, webp_file_path, is_debug):
        # whether has already converted
        if exists(webp_file_path):
            # has already converted!
            _reduce_size = size_diff(image_file_path, webp_file_path)

            if _reduce_size < 0:
                remove(webp_file_path)
                print_process(
                    image_file_name + ' has already converted, but it larger than origin-image: ' + _reduce_size.__str__() + ' so, remove it and re-convert it!')
            else:
                print_process(image_file_name + ' has already converted! reduce: ' + _reduce_size.__str__())
                return RESULT_ALREADY_EXIST, _reduce_size

        # transparency
        if ignore_transparency_img and image_file_name.endswith('.png'):
            from PIL import Image

            try:
                img = Image.open(image_file_path, 'r')
            except IOError:
                print_process("NOT convert " + image_file_name + ' because convert failed!')
                return RESULT_FAILED, 0

            if img.mode in ('RGBA', 'LA') or (img.mode == 'P' and 'transparency' in img.info):
                alpha = img.convert('RGBA').split()[-1]
                for pixel in alpha.getdata():
                    if pixel < 255:
                        print_process(
                            'NOT convert ' + image_file_name + ' because there is alpha[' + pixel.__str__() + ']')
                        return RESULT_WITH_TRANSPARENCY, 0

        print_process('convert for ' + image_file_path)

        swap_webp_path = self.swap_webp_path

        command = list(self.command_prefix)
        command.append(image_file_path)
        command.extend(['-o', swap_webp_path])

        if is_debug:
            subprocess.Popen(command).communicate()
            print(command.__str__())
        else:
            subprocess.Popen(command, stdout=FNULL, stderr=subprocess.STDOUT).communicate()

        # print command_prefix + image_file_path + ' -o ' + swap_webp_path
        # system(command_prefix + image_file_path + ' -o ' + swap_webp_path)
        # whether convert failed
        if not exists(swap_webp_path):
            # convert failed!
            print_process("NOT convert " + image_file_name + ' because convert failed!')
            return RESULT_FAILED, 0

        _reduce_size = size_diff(image_file_path, swap_webp_path)
        if _reduce_size <= 0:
            # invalid convert
            print_process(
                'NOT convert ' + image_file_name + ' because the webp one is larger: ' + (-_reduce_size).__str__())
            remove(swap_webp_path)
            return RESULT_WEBP_LARGER, 0

        print_process('convert ' + image_file_name + ' and reduce size: ' + _reduce_size.__str__())
        rename(swap_webp_path, webp_file_path)
        if exists(swap_webp_path): remove(swap_webp_path)
        return RESULT_SUCCESS, _reduce_size


webp_converter.py
#!/usr/bin/python -u

"""
Copyright 2017, JacksGong(https://blog.dreamtobe.cn)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

import os
import re
import sys
import time
from os import makedirs, remove, rename
from os.path import getsize, exists, join
from shutil import copyfile, rmtree
from sys import argv, exit

__version__ = '3.0.0'
__author__ = 'JacksGong'

RESET = '\033[0m'
BLACK, RED, GREEN, YELLOW, BLUE, MAGENTA, CYAN, WHITE = range(8)


def termcolor(fg=None, bg=None):
    codes = []
    if fg is not None: codes.append('3%d' % fg)
    if bg is not None: codes.append('10%d' % bg)
    return '\033[%sm' % ';'.join(codes) if codes else ''


def colorize(message, fg=None, bg=None):
    return termcolor(fg, bg) + message + RESET


def print_process(message):
    print colorize(message, fg=YELLOW)


def size_diff(left, right):
    return getsize(left) - getsize(right)


def human_bytes(B):
    """Return the given bytes as a human friendly KB, MB, GB, or TB string"""
    B = float(B)
    KB = float(1024)
    MB = float(KB ** 2)  # 1,048,576
    GB = float(KB ** 3)  # 1,073,741,824
    TB = float(KB ** 4)  # 1,099,511,627,776

    if B < KB:
        return '{0} {1}'.format(B, 'Bytes' if 0 == B > 1 else 'Byte')
    elif KB <= B < MB:
        return '{0:.2f} KB'.format(B / KB)
    elif MB <= B < GB:
        return '{0:.2f} MB'.format(B / MB)
    elif GB <= B < TB:
        return '{0:.2f} GB'.format(B / GB)
    elif TB <= B:
        return '{0:.2f} TB'.format(B / TB)


def resource_path(relative_path):
    # noinspection PyProtectedMember
    return join(sys._MEIPASS, relative_path) if hasattr(sys, '_MEIPASS') else relative_path


def convert(image_file_path, image_file_name, webp_file_path):
    # whether has already converted
    if exists(webp_file_path):
        # has already converted!
        _reduce_size = size_diff(image_file_path, webp_file_path)

        if _reduce_size < 0:
            remove(webp_file_path)
            print_process(
                image_file_name + ' has already converted, but it larger than origin-image: ' + _reduce_size.__str__() + ' so, remove it and re-convert it!')
        else:
            print_process(image_file_name + ' has already converted! reduce: ' + _reduce_size.__str__())
            return RESULT_ALREADY_EXIST, _reduce_size

    # transparency
    if ignore_transparency_image and image_file_name.endswith('.png'):
        from PIL import Image

        try:
            img = Image.open(image_file_path, 'r')
        except IOError:
            print_process("NOT convert " + image_file_name + ' because convert failed!')
            return RESULT_FAILED, 0

        if img.mode in ('RGBA', 'LA') or (img.mode == 'P' and 'transparency' in img.info):
            alpha = img.convert('RGBA').split()[-1]
            for pixel in alpha.getdata():
                if pixel < 255:
                    print_process('NOT convert ' + image_file_name + ' because there is alpha[' + pixel.__str__() + ']')
                    return RESULT_WITH_TRANSPARENCY, 0

    print_process('convert for ' + image_file_name)

    os.system(command_prefix + image_file_path + ' -o ' + swap_webp_path)
    # whether convert failed
    if not exists(swap_webp_path):
        # convert failed!
        print_process("NOT convert " + image_file_name + ' because convert failed!')
        return RESULT_FAILED, 0

    _reduce_size = size_diff(image_file_path, swap_webp_path)
    if _reduce_size < 0:
        # invalid convert
        print_process(
            'NOT convert ' + image_file_name + ' because the webp one is larger: ' + (-_reduce_size).__str__())
        remove(swap_webp_path)
        return RESULT_WEBP_LARGER, 0

    print_process('convert ' + image_file_name + ' and reduce size: ' + _reduce_size.__str__())
    rename(swap_webp_path, webp_file_path)
    if exists(swap_webp_path): remove(swap_webp_path)
    return RESULT_SUCCESS, _reduce_size


# ==================================
start_time = time.time()

image_dir_path = None
quality_ratio = 100
ignore_transparency_image = False
replace = False

# load config
image_dir_path_re = re.compile(r'image-path: *(.*)')
quality_ratio_re = re.compile(r'quality-ratio: *(\d*)')
ignore_transparency_re = re.compile(r'ignore-transparency-image: *(.*)')
replace_re = re.compile(r'replace: *(.*)')

conf_file_path = '.webp.conf'
if not exists(conf_file_path):
    exit(colorize(
        'Please create \'.webp.conf\' file, and add config:\nimage-path: /the/origin/image/path\nquality-ratio:[0~100](default is 100)\nignore-transparency-image: [true/false]\nreplace: [true/false]',
        fg=RED))

conf_file = open(conf_file_path, 'r')
for line in conf_file:
    line = line.strip()
    if line.startswith('#') or line.startswith('/') or line.startswith('*'):
        continue
    image_dir_path_process = image_dir_path_re.search(line)
    if image_dir_path_process is not None:
        image_dir_path = image_dir_path_process.groups()[0]
        continue

    quality_ratio_process = quality_ratio_re.search(line)
    if quality_ratio_process is not None:
        quality_ratio = int(quality_ratio_process.groups()[0])
        continue

    ignore_transparency_process = ignore_transparency_re.search(line)
    if ignore_transparency_process is not None:
        ignore_transparency_image = ignore_transparency_process.groups()[0] == 'true'
        continue

    replace_process = replace_re.search(line)
    if replace_process is not None:
        replace = replace_process.groups()[0] == 'true'
        continue

    exit(colorize('unknown config ' + line, fg=YELLOW))
conf_file.close()

if image_dir_path is None:
    exit(colorize(
        'Please create add config:\nimage-path: /the/origin/image/path', fg=RED))

command_prefix = '%s ' % resource_path('cwebp')
if quality_ratio >= 100:
    command_prefix += '-lossless -q 100 '
else:
    command_prefix += '-q ' + quality_ratio.__str__() + ' '

is_clean_env = False
if argv.__len__() > 1:
    value = argv[1]
    if value == '-withClean':
        if replace:
            exit(colorize(
                "can't clean the origin environment: can't withClean for replace mode, please remove 'replace' profile on .webp.conf or not execute with '-withClean' param",
                fg=RED))
        is_clean_env = True
    else:
        exit(colorize(
            "unknown arg: " + value + " only support '-withClean' to clean environment", fg=RED))

root = 'webp-converted/'
keep_origin_path = root + 'origin/'
convert_fail_path = root + 'failed/'
transparency_image_path = root + 'transparency/'
swap_webp_path = root + 'swap.webp'

if not exists(root):
    makedirs(root)
elif is_clean_env:
    rmtree(root)

all_reduce_size = 0
valid_convert_file_count = 0
failed_convert_count = 0
scan_file_count = 0
skip_file_count = 0
skip_transparency_file_count = 0

if exists(swap_webp_path):
    remove(swap_webp_path)

RESULT_SUCCESS = 0
RESULT_FAILED = -1
RESULT_WEBP_LARGER = -2
RESULT_WITH_TRANSPARENCY = -3
RESULT_ALREADY_EXIST = -4

for subdir, dirs, files in os.walk(image_dir_path):
    for file_name in files:
        if file_name == '.DS_Store':
            continue
        if not file_name.endswith('.jpg') and not file_name.endswith('.png'):
            continue

        scan_file_count += 1
        input_file_path = join(subdir, file_name)
        output_file_name = file_name.rsplit('.', 1)[0] + '.webp'
        if replace:
            output_file_path = join(subdir, output_file_name)
        else:
            output_file_path = join(root, output_file_name)

        result, reduce_size = convert(input_file_path, file_name, output_file_path)
        if reduce_size > 0:
            all_reduce_size += reduce_size
            if result != RESULT_ALREADY_EXIST:
                valid_convert_file_count += 1
        elif result == RESULT_FAILED:
            failed_convert_count += 1
        elif result == RESULT_WEBP_LARGER:
            skip_file_count += 1
        elif result == RESULT_WITH_TRANSPARENCY:
            skip_transparency_file_count += 1

        if replace:
            if reduce_size > 0:
                remove(input_file_path)
        else:
            if result == RESULT_WITH_TRANSPARENCY:
                if not exists(transparency_image_path):
                    makedirs(transparency_image_path)
                copyfile(input_file_path, transparency_image_path + file_name)
            elif result == RESULT_FAILED:
                if not exists(convert_fail_path):
                    makedirs(convert_fail_path)
                copyfile(input_file_path, convert_fail_path + file_name)
            elif result == RESULT_WEBP_LARGER:
                if not exists(keep_origin_path):
                    makedirs(keep_origin_path)
                copyfile(input_file_path, keep_origin_path + file_name)

print '-----------------------------------------------'
print ' '
if replace:
    print colorize('Replace ' + valid_convert_file_count.__str__() + ' image files on ' + image_dir_path, fg=CYAN)
else:
    print colorize('All files handled on: ', fg=CYAN) + root

print colorize('Consume: ', fg=CYAN) + '%ss' % (time.time() - start_time)
print ' '
print colorize('Scan files count: ', fg=GREEN) + scan_file_count.__str__()
print colorize('Converted files count: ', fg=GREEN) + valid_convert_file_count.__str__()
print colorize('Reduce size: ', fg=GREEN) + human_bytes(all_reduce_size)
print colorize('Skip files(because convert failed) count: ',
               fg=GREEN) + failed_convert_count.__str__()
print colorize('Skip files(because the webp one is greater than origin one) count: ',
               fg=GREEN) + skip_file_count.__str__()
if skip_transparency_file_count > 0:
    print colorize('Skip files(because there is transparency) count: ',
                   fg=GREEN) + skip_transparency_file_count.__str__()
print ' '
print '-----------------------------------------------'
